src/extensions/timeseries/TimeseriesViewNew/useTimeseriesModel.ts

```diff
@@ -1,229 +1,161 @@
   }
-  getChannelData(ch: number, t1: number, t2: number, ds_factor: number): number[] {
-    const segment_size = this.args.segment_size
-    const s1 = Math.floor(t1 / segment_size);
-    const s2 = Math.floor(t2 / segment_size);
-    this.requestChannelData(ch, t1, t2, ds_factor)
+  let X = new Mda()
+  X.setFromBase64(result.data_b64);
+  return X
+}
 
-    const ret = [];
-    if (ds_factor === 1) {
-      for (let t = t1; t < t2; t++) {
-        ret.push(NaN);
+type TimeseriesDataSegmentQuery = {
+  type: 'dataSegment',
+  ds_factor: number,
+  segment_num: number,
+  segment_size: number
+}
+
+type TimeseriesDataQuery = TimeseriesDataSegmentQuery
+
+const useTimeseriesData = (recordingObject: any, recordingInfo: RecordingInfo): TimeseriesData | null => {
+  const hither = useContext(HitherContext)
+  const fetch = useMemo(() => (async (query: TimeseriesDataQuery) => {
+    switch(query.type) {
+      case 'dataSegment': {
+        return await getTimeseriesDataSegment({hither, recordingObject, ds_factor: query.ds_factor, segment_num: query.segment_num, segment_size: query.segment_size})
       }
     }
-    else {
-      for (let t = t1; t < t2; t++) {
-        ret.push(NaN);
-        ret.push(NaN);
+  }), [hither, recordingObject])
+  const data = useFetchCache<TimeseriesDataQuery>(fetch)
+
+  const segment_size_times_num_channels = 100000
+  const num_channels = recordingInfo.channel_ids.length
+  const segment_size = Math.ceil(segment_size_times_num_channels / num_channels)
+
+  const getChannelData = useMemo(() => ((ch: number, t1: number, t2: number, ds_factor: number) => {
+    const i1 = Math.floor(t1 / segment_size)
+    const i2 = Math.ceil((t2 -1) / segment_size)
+    const segments = []
+    for (let i = i1; i <= i2; i++) {
+      if ((i === i1) && (i === i2)) {
+        segments.push({
+          segment_num: i,
+          t1: i * segment_size,
+          t2: (i + 1) * segment_size,
+          src1: t1 - (i * segment_size),
+          src2: t2 - (i * segment_size),
+          dst1: 0,
+          dst2: t2 - t1
+        })
+      }
+      else if (i === i1) {
+        segments.push({
+          segment_num: i,
+          t1: i * segment_size,
+          t2: (i + 1) * segment_size,
+          src1: t1 - (i * segment_size),
+          src2: segment_size,
+          dst1: 0,
+          dst2: segment_size + (i * segment_size) - t1
+        })
+      }
+      else if (i === i2) {
+        segments.push({
+          segment_num: i,
+          t1: i * segment_size,
+          t2: (i + 1) * segment_size,
+          src1: 0,
+          src2: t2 - (i * segment_size),
+          dst1: i * segment_size - t1,
+          dst2: t2 - t1
+        })
+      }
+      else {
+        segments.push({
+          segment_num: i,
+          t1: i * segment_size,
+          t2: (i + 1) * segment_size,
+          src1: 0,
+          src2: segment_size,
+          dst1: i * segment_size - t1,
+          dst2: (i + 1) * segment_size - t1
+        })
```

Comment:
I understand this is how to fetch different segments based on where we are in the time series (yes?) but the subtleties of the mechanics are escaping me right now. It's also not immediately clear what `src` and `dst` are referring to. I think some comments would be helpful here.

Reply:
I agree, I'll create an issue for this.